### ISSUE TITLE
Retrieving free disk space on Windows can time out

### ISSUE THREAD

**Issue by lukebakken** [rabbitmq/rabbitmq-server#4140]
Reported here -

https://groups.google.com/g/rabbitmq-users/c/WwRMEs08MOo

Related to -
#3895
#3970
#3974

**Comment by phiamo** [comment]
Hi @lukebakken 
I have a similar issue  on ubuntu linux:

 Command timed out /usr/bin/df -kP /var/lib/rabbitmq/mnesia/rabbit@[redacted-node-name]

i happened regularly once a day, until i upgraded erlang etc from https://packages.erlang-solutions.com/ubuntu
no it happenens every hour :/

Main pitfall, after this happens the server never recovers, and monitoring tells me rabbit is down, until i manually restart it

```
2022-02-24 12:56:07.744075+00:00 [error] <0.1118.0> missed heartbeats from client, timeout: 10s
2022-02-24 12:56:07.749075+00:00 [warning] <0.1517.0> closing AMQP connection <0.1517.0> (54.93.233.76:59614 -> 172.31.4.249:5671, vhost: '[redacted-node-name]', user: 'customer_consumer'):
2022-02-24 12:56:07.749075+00:00 [warning] <0.1517.0> client unexpectedly closed TCP connection
2022-02-24 12:55:51.653764+00:00 [error] <0.184.0> ** Generic server aten_detector terminating
2022-02-24 12:55:51.653764+00:00 [error] <0.184.0> ** Last message in was poll
2022-02-24 12:55:51.653764+00:00 [error] <0.184.0> ** When Server state == {state,#Ref<0.859878212.2858156034.85769>,5000,0.99,
2022-02-24 12:55:51.653764+00:00 [error] <0.184.0>                                #{},#{}}
2022-02-24 12:55:51.653764+00:00 [error] <0.184.0> ** Reason for termination ==
2022-02-24 12:55:51.653764+00:00 [error] <0.184.0> ** {{timeout,{gen_server,call,[aten_sink,get_failure_probabilities]}},
2022-02-24 12:55:51.653764+00:00 [error] <0.184.0>     [{gen_server,call,2,[{file,"gen_server.erl"},{line,239}]},
2022-02-24 12:55:51.653764+00:00 [error] <0.184.0>      {aten_detector,handle_info,2,[{file,"src/aten_detector.erl"},{line,109}]},
2022-02-24 12:55:51.653764+00:00 [error] <0.184.0>      {gen_server,try_dispatch,4,[{file,"gen_server.erl"},{line,695}]},
2022-02-24 12:55:51.653764+00:00 [error] <0.184.0>      {gen_server,handle_msg,6,[{file,"gen_server.erl"},{line,771}]},
2022-02-24 12:55:51.653764+00:00 [error] <0.184.0>      {proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,226}]}]}
2022-02-24 12:55:51.653764+00:00 [error] <0.184.0> 
2022-02-24 12:56:55.130672+00:00 [warning] <0.1190.0> closing AMQP connection <0.1190.0> (18.159.106.4:42084 -> 172.31.4.249:5671, vhost: '[redacted-node-name]', user: 'file_storage_consumer'):
2022-02-24 12:56:55.130672+00:00 [warning] <0.1190.0> client unexpectedly closed TCP connection
2022-02-24 12:57:11.028232+00:00 [error] <0.1303.0> closing AMQP connection <0.1303.0> (18.159.106.4:48640 -> 172.31.4.249:5671):
2022-02-24 12:57:11.028232+00:00 [error] <0.1303.0> missed heartbeats from client, timeout: 20s
2022-02-24 12:56:39.365290+00:00 [error] <0.184.0>   crasher:
2022-02-24 12:56:39.365290+00:00 [error] <0.184.0>     initial call: aten_detector:init/1
2022-02-24 12:56:39.365290+00:00 [error] <0.184.0>     pid: <0.184.0>
2022-02-24 12:56:39.365290+00:00 [error] <0.184.0>     registered_name: aten_detector
2022-02-24 12:56:39.365290+00:00 [error] <0.184.0>     exception exit: {timeout,
2022-02-24 12:56:39.365290+00:00 [error] <0.184.0>                         {gen_server,call,
2022-02-24 12:56:39.365290+00:00 [error] <0.184.0>                             [aten_sink,get_failure_probabilities]}}
2022-02-24 12:56:39.365290+00:00 [error] <0.184.0>       in function  gen_server:call/2 (gen_server.erl, line 239)
2022-02-24 12:56:39.365290+00:00 [error] <0.184.0>       in function  gen_server:call/2 (gen_server.erl, line 239)
2022-02-24 12:56:39.365290+00:00 [error] <0.184.0>       in call from aten_detector:handle_info/2 (src/aten_detector.erl, line 109)
2022-02-24 12:56:39.365290+00:00 [error] <0.184.0>       in call from gen_server:try_dispatch/4 (gen_server.erl, line 695)
2022-02-24 12:56:39.365290+00:00 [error] <0.184.0>       in call from gen_server:handle_msg/6 (gen_server.erl, line 771)
2022-02-24 12:56:39.365290+00:00 [error] <0.184.0>     ancestors: [aten_sup,<0.180.0>]
2022-02-24 12:56:39.365290+00:00 [error] <0.184.0>     message_queue_len: 1
2022-02-24 12:56:39.365290+00:00 [error] <0.184.0>     messages: [poll]
2022-02-24 12:56:39.365290+00:00 [error] <0.184.0>     links: [<0.181.0>]
2022-02-24 12:56:39.365290+00:00 [error] <0.184.0>     dictionary: []
2022-02-24 12:56:39.365290+00:00 [error] <0.184.0>     trap_exit: false
2022-02-24 12:56:39.365290+00:00 [error] <0.184.0>     status: running
2022-02-24 12:56:39.365290+00:00 [error] <0.184.0>     heap_size: 6772
2022-02-24 12:56:39.365290+00:00 [error] <0.184.0>     stack_size: 29
2022-02-24 12:56:39.365290+00:00 [error] <0.184.0>     reductions: 97632
2022-02-24 12:56:39.365290+00:00 [error] <0.184.0>   neighbours:
2022-02-24 12:56:39.365290+00:00 [error] <0.184.0> 
2022-02-24 12:58:13.861539+00:00 [error] <0.1560.0> closing AMQP connection <0.1560.0> (54.93.233.76:59616 -> 172.31.4.249:5671):
2022-02-24 12:58:13.861539+00:00 [error] <0.1560.0> missed heartbeats from client, timeout: 20s
2022-02-24 12:58:13.720459+00:00 [error] <0.181.0>     supervisor: {local,aten_sup}
2022-02-24 12:58:13.720459+00:00 [error] <0.181.0>     errorContext: child_terminated
2022-02-24 12:58:13.720459+00:00 [error] <0.181.0>     reason: {timeout,{gen_server,call,[aten_sink,get_failure_probabilities]}}
2022-02-24 12:58:13.720459+00:00 [error] <0.181.0>     offender: [{pid,<0.184.0>},
2022-02-24 12:58:13.720459+00:00 [error] <0.181.0>                {id,aten_detector},
2022-02-24 12:58:13.720459+00:00 [error] <0.181.0>                {mfargs,{aten_detector,start_link,[]}},
2022-02-24 12:58:13.720459+00:00 [error] <0.181.0>                {restart_type,permanent},
2022-02-24 12:58:13.720459+00:00 [error] <0.181.0>                {significant,false},
2022-02-24 12:58:13.720459+00:00 [error] <0.181.0>                {shutdown,5000},
2022-02-24 12:58:13.720459+00:00 [error] <0.181.0>                {child_type,worker}]
2022-02-24 12:58:13.720459+00:00 [error] <0.181.0> 
2022-02-24 12:58:45.416636+00:00 [error] <0.376.0> Command timed out: '/usr/bin/df -kP /var/lib/rabbitmq/mnesia/rabbit@[redacted-node-name]'
2022-02-24 12:58:45.416752+00:00 [error] <0.376.0> ** Generic server rabbit_disk_monitor terminating
2022-02-24 12:58:45.416752+00:00 [error] <0.376.0> ** Last message in was update
2022-02-24 12:58:45.416752+00:00 [error] <0.376.0> ** When Server state == {state,"/var/lib/rabbitmq/mnesia/rabbit@[redacted-node-name]",
2022-02-24 12:58:45.416752+00:00 [error] <0.376.0>                                50000000,46926032896,100,10000,
2022-02-24 12:58:45.416752+00:00 [error] <0.376.0>                                #Ref<0.859878212.2858156033.2126>,false,true,
2022-02-24 12:58:45.416752+00:00 [error] <0.376.0>                                10,120000}
2022-02-24 12:58:45.416752+00:00 [error] <0.376.0> ** Reason for termination ==
2022-02-24 12:58:45.416752+00:00 [error] <0.376.0> ** {function_clause,
2022-02-24 12:58:45.416752+00:00 [error] <0.376.0>        [{lists,reverse,[{error,timeout}],[{file,"lists.erl"},{line,147}]},
2022-02-24 12:58:45.416752+00:00 [error] <0.376.0>         {string,tokens,2,[{file,"string.erl"},{line,1934}]},
2022-02-24 12:58:45.416752+00:00 [error] <0.376.0>         {rabbit_disk_monitor,parse_free_unix,1,
2022-02-24 12:58:45.416752+00:00 [error] <0.376.0>             [{file,"rabbit_disk_monitor.erl"},{line,262}]},
2022-02-24 12:58:45.416752+00:00 [error] <0.376.0>         {rabbit_disk_monitor,internal_update,1,
2022-02-24 12:58:45.416752+00:00 [error] <0.376.0>             [{file,"rabbit_disk_monitor.erl"},{line,216}]},
2022-02-24 12:58:45.416752+00:00 [error] <0.376.0>         {rabbit_disk_monitor,handle_info,2,
2022-02-24 12:58:45.416752+00:00 [error] <0.376.0>             [{file,"rabbit_disk_monitor.erl"},{line,166}]},
2022-02-24 12:58:45.416752+00:00 [error] <0.376.0>         {gen_server,try_dispatch,4,[{file,"gen_server.erl"},{line,695}]},
2022-02-24 12:58:45.416752+00:00 [error] <0.376.0>         {gen_server,handle_msg,6,[{file,"gen_server.erl"},{line,771}]},
2022-02-24 12:58:45.416752+00:00 [error] <0.376.0>         {proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,226}]}]}
2022-02-24 12:58:45.416752+00:00 [error] <0.376.0> 
```

**Comment by lukebakken** [comment]
@phiamo your issue is not the same as this github issue. Please start a discussion via this link:

https://github.com/rabbitmq/rabbitmq-server/discussions

*ATTACH* your log information as a file, do not paste it.

I will delete your comment and this one later today.

**Comment by itmay** [comment]
I have the same issue. Logs: [rabbitmq_powershell_timeout.txt](https://github.com/rabbitmq/rabbitmq-server/files/8221995/rabbitmq_powershell_timeout.txt)

I observed this after losing metrics in prometheus. I got error: `strconv.ParseFloat: parsing "unknown": invalid syntax`.

In metrics I have:
```
# TYPE rabbitmq_disk_space_available_bytes gauge
# HELP rabbitmq_disk_space_available_bytes Disk space available in bytes
rabbitmq_disk_space_available_bytes unknown


```

**Comment by lukebakken** [comment]
@itmay - not being able to retrieve free disk space from Powershell is a _symptom_ of an overloaded system.

**Comment by michaelklishin** [comment]
Unlike #4751 where I don't know what realistically can be improved (a `df` call isn't clever or resource-intensive design by any stretch), there is hope that #4328 can make free disk space calculation on Windows significantly more efficient.

**Comment by querqueq** [comment]
Have you considered getting the free disk space via the win32 API with a NIF? I am not familiar with Erlang or BEAM and the problem with NIFs seems to be that they can crash BEAM. 
There's a rust library called [rustler](https://github.com/rusterlium/rustler) which supposedly can create NIFs which will not crash BEAM.
There are also Microsoft maintained bindings to win32 API, [windows-rs](https://github.com/microsoft/windows-rs), which could be used to get the free disk space with `GetDiskFreeSpaceExW`.

MSDN: https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-getdiskfreespaceexw
Crate Docs: https://microsoft.github.io/windows-docs-rs/doc/windows/Win32/Storage/FileSystem/fn.GetDiskFreeSpaceExW.html

So free disk space could be retrieved with windows-rs via the win32 API and a safe NIF could be created with rustler which then can be used in RMQ.

**Comment by lukebakken** [comment]
> Have you considered getting the free disk space via the win32 API with a NIF?

Yep, we sure have. The main issue is that building Erlang and NIFs on Windows isn't as trivial as on a POSIX system.

At some point I'm going to return to #4328.

**Comment by michaelklishin** [comment]
We also have considered developing an open source equivalent of of `handle.exe` but it's not easy to justify the time investment.

**Comment by baoanh194** [comment]
Hi everyone,

Is there any update on the free disk space issue?

**Comment by sturla78** [comment]
Hello,
i have same issue, on some windows virtual machine the command Get-PSDrive take a lot of seconds and the cpu is hight, on the rabbit log there is the following error:

(RabbitMQ 3.9.19 on Erlang 23.3.4.14)

```
2022-07-05 11:12:04.998000+02:00 [error] <0.400.0> Command timed out: 'powershell.exe -NoLogo -NoProfile -NonInteractive -Command (Get-PSDrive c).Free'
2022-07-05 11:12:04.998000+02:00 [error] <0.400.0> ** Generic server rabbit_disk_monitor terminating
2022-07-05 11:12:04.998000+02:00 [error] <0.400.0> ** Last message in was update
2022-07-05 11:12:04.998000+02:00 [error] <0.400.0> ** When Server state == {state,"c:/Users/Administrator/AppData/Roaming/RabbitMQ/db/rabbit@vmErlTest-mnesia",
2022-07-05 11:12:04.998000+02:00 [error] <0.400.0>                                50000000,60257439744,100,10000,
2022-07-05 11:12:04.998000+02:00 [error] <0.400.0>                                #Ref<0.2713404489.916193282.77281>,false,true,
2022-07-05 11:12:04.998000+02:00 [error] <0.400.0>                                10,120000}
2022-07-05 11:12:04.998000+02:00 [error] <0.400.0> ** Reason for termination ==
2022-07-05 11:12:04.998000+02:00 [error] <0.400.0> ** {could_not_determine_disk_free,
2022-07-05 11:12:04.998000+02:00 [error] <0.400.0>        [{rabbit_disk_monitor,get_disk_free,2,
2022-07-05 11:12:04.998000+02:00 [error] <0.400.0>             [{file,"rabbit_disk_monitor.erl"},{line,257}]},
2022-07-05 11:12:04.998000+02:00 [error] <0.400.0>         {rabbit_disk_monitor,internal_update,1,
2022-07-05 11:12:04.998000+02:00 [error] <0.400.0>             [{file,"rabbit_disk_monitor.erl"},{line,216}]},
2022-07-05 11:12:04.998000+02:00 [error] <0.400.0>         {rabbit_disk_monitor,handle_info,2,
2022-07-05 11:12:04.998000+02:00 [error] <0.400.0>             [{file,"rabbit_disk_monitor.erl"},{line,166}]},
2022-07-05 11:12:04.998000+02:00 [error] <0.400.0>         {gen_server,try_dispatch,4,[{file,"gen_server.erl"},{line,689}]},
2022-07-05 11:12:04.998000+02:00 [error] <0.400.0>         {gen_server,handle_msg,6,[{file,"gen_server.erl"},{line,765}]},
2022-07-05 11:12:04.998000+02:00 [error] <0.400.0>         {proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,226}]}]}
2022-07-05 11:12:04.998000+02:00 [error] <0.400.0> 
2022-07-05 11:12:04.999000+02:00 [error] <0.400.0>   crasher:
2022-07-05 11:12:04.999000+02:00 [error] <0.400.0>     initial call: rabbit_disk_monitor:init/1
2022-07-05 11:12:04.999000+02:00 [error] <0.400.0>     pid: <0.400.0>
2022-07-05 11:12:04.999000+02:00 [error] <0.400.0>     registered_name: rabbit_disk_monitor
2022-07-05 11:12:04.999000+02:00 [error] <0.400.0>     exception exit: could_not_determine_disk_free
2022-07-05 11:12:04.999000+02:00 [error] <0.400.0>       in function  rabbit_disk_monitor:get_disk_free/2 (rabbit_disk_monitor.erl, line 257)
2022-07-05 11:12:04.999000+02:00 [error] <0.400.0>       in call from rabbit_disk_monitor:internal_update/1 (rabbit_disk_monitor.erl, line 216)
2022-07-05 11:12:04.999000+02:00 [error] <0.400.0>       in call from rabbit_disk_monitor:handle_info/2 (rabbit_disk_monitor.erl, line 166)
2022-07-05 11:12:04.999000+02:00 [error] <0.400.0>       in call from gen_server:try_dispatch/4 (gen_server.erl, line 689)
2022-07-05 11:12:04.999000+02:00 [error] <0.400.0>       in call from gen_server:handle_msg/6 (gen_server.erl, line 765)
2022-07-05 11:12:04.999000+02:00 [error] <0.400.0>     ancestors: [rabbit_disk_monitor_sup,rabbit_sup,<0.230.0>]
2022-07-05 11:12:04.999000+02:00 [error] <0.400.0>     message_queue_len: 0
2022-07-05 11:12:04.999000+02:00 [error] <0.400.0>     messages: []
2022-07-05 11:12:04.999000+02:00 [error] <0.400.0>     links: [<0.399.0>]
2022-07-05 11:12:04.999000+02:00 [error] <0.400.0>     dictionary: []
2022-07-05 11:12:04.999000+02:00 [error] <0.400.0>     trap_exit: false
2022-07-05 11:12:04.999000+02:00 [error] <0.400.0>     status: running
2022-07-05 11:12:04.999000+02:00 [error] <0.400.0>     heap_size: 6772
2022-07-05 11:12:04.999000+02:00 [error] <0.400.0>     stack_size: 28
2022-07-05 11:12:04.999000+02:00 [error] <0.400.0>     reductions: 119819
2022-07-05 11:12:04.999000+02:00 [error] <0.400.0>   neighbours:
2022-07-05 11:12:04.999000+02:00 [error] <0.400.0> 
2022-07-05 11:12:05.000000+02:00 [error] <0.399.0>     supervisor: {local,rabbit_disk_monitor_sup}
2022-07-05 11:12:05.000000+02:00 [error] <0.399.0>     errorContext: child_terminated
2022-07-05 11:12:05.000000+02:00 [error] <0.399.0>     reason: could_not_determine_disk_free
2022-07-05 11:12:05.000000+02:00 [error] <0.399.0>     offender: [{pid,<0.400.0>},
2022-07-05 11:12:05.000000+02:00 [error] <0.399.0>                {id,rabbit_disk_monitor},
2022-07-05 11:12:05.000000+02:00 [error] <0.399.0>                {mfargs,{rabbit_disk_monitor,start_link,[50000000]}},
2022-07-05 11:12:05.000000+02:00 [error] <0.399.0>                {restart_type,{transient,1}},
2022-07-05 11:12:05.000000+02:00 [error] <0.399.0>                {shutdown,300000},
2022-07-05 11:12:05.000000+02:00 [error] <0.399.0>                {child_type,worker}]
```

**Comment by michaelklishin** [comment]
@baoanh194 there are no updates, if an issue is addressed, it would be closed. This is open source software, so you are welcome to dive in and contribute a solution if you badly need one.

We have outlined two possible options above, one is https://github.com/rabbitmq/rabbitmq-server/pull/4328 and another is developing an open source `handle.exe` alternative we can target and recommend eventually.

**Comment by noobwu** [comment]
Hello,
i have same issue, on  windows(Windows Server 2022 Datacenter 21H2 ) virtual machine 

(RabbitMQ 3.10.6 on Erlang 25.0.3)
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1> Command timed out: 'powershell.exe -NoLogo -NoProfile -NonInteractive -Command (Get-PSDrive c).Free'
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1> ** Generic server rabbit_disk_monitor terminating
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1> ** Last message in was try_enable
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1> ** When Server state == {state,"c:/Users/xtt~aly/AppData/Roaming/RabbitMQ/db/rabbit@iZq8zm58oi8hwrZ-mnesia",
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>                                50000000,undefined,100,10000,undefined,false,
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>                                false,9,120000}
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1> ** Reason for termination ==
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1> ** {could_not_determine_disk_free,
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>        [{rabbit_disk_monitor,get_disk_free,2,
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>             [{file,"rabbit_disk_monitor.erl"},{line,254}]},
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>         {rabbit_disk_monitor,internal_update,1,
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>             [{file,"rabbit_disk_monitor.erl"},{line,216}]},
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>         {rabbit_disk_monitor,enable,1,
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>             [{file,"rabbit_disk_monitor.erl"},{line,355}]},
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>         {rabbit_disk_monitor,handle_info,2,
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>             [{file,"rabbit_disk_monitor.erl"},{line,164}]},
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>         {gen_server,try_dispatch,4,[{file,"gen_server.erl"},{line,1120}]},
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>         {gen_server,handle_msg,6,[{file,"gen_server.erl"},{line,1197}]},
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>         {proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,240}]}]}
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1> 
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>   crasher:
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>     initial call: rabbit_disk_monitor:init/1
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>     pid: <0.5482.1>
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>     registered_name: rabbit_disk_monitor
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>     exception exit: could_not_determine_disk_free
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>       in function  rabbit_disk_monitor:get_disk_free/2 (rabbit_disk_monitor.erl, line 254)
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>       in call from rabbit_disk_monitor:internal_update/1 (rabbit_disk_monitor.erl, line 216)
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>       in call from rabbit_disk_monitor:enable/1 (rabbit_disk_monitor.erl, line 355)
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>       in call from rabbit_disk_monitor:handle_info/2 (rabbit_disk_monitor.erl, line 164)
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>       in call from gen_server:try_dispatch/4 (gen_server.erl, line 1120)
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>       in call from gen_server:handle_msg/6 (gen_server.erl, line 1197)
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>     ancestors: [rabbit_disk_monitor_sup,rabbit_sup,<0.219.0>]
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>     message_queue_len: 0
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>     messages: []
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>     links: [<0.354.0>]
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>     dictionary: []
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>     trap_exit: false
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>     status: running
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>     heap_size: 6772
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>     stack_size: 28
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>     reductions: 33064
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>   neighbours:
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1> 
2022-08-16 12:26:52.084000+08:00 [error] <0.354.0>     supervisor: {local,rabbit_disk_monitor_sup}
2022-08-16 12:26:52.084000+08:00 [error] <0.354.0>     errorContext: child_terminated
2022-08-16 12:26:52.084000+08:00 [error] <0.354.0>     reason: could_not_determine_disk_free
2022-08-16 12:26:52.084000+08:00 [error] <0.354.0>     offender: [{pid,<0.5482.1>},
2022-08-16 12:26:52.084000+08:00 [error] <0.354.0>                {id,rabbit_disk_monitor},
2022-08-16 12:26:52.084000+08:00 [error] <0.354.0>                {mfargs,{rabbit_disk_monitor,start_link,[50000000]}},
2022-08-16 12:26:52.084000+08:00 [error] <0.354.0>                {restart_type,{transient,1}},
2022-08-16 12:26:52.084000+08:00 [error] <0.354.0>                {shutdown,300000},
2022-08-16 12:26:52.084000+08:00 [error] <0.354.0>                {child_type,worker}]

**Comment by noobwu** [comment]
Hello,
i have same issue, on  windows(Windows Server 2022 Datacenter 21H2 ) virtual machine 

(RabbitMQ 3.10.6 on Erlang 25.0.3)
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1> Command timed out: 'powershell.exe -NoLogo -NoProfile -NonInteractive -Command (Get-PSDrive c).Free'
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1> ** Generic server rabbit_disk_monitor terminating
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1> ** Last message in was try_enable
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1> ** When Server state == {state,"c:/Users/xtt~aly/AppData/Roaming/RabbitMQ/db/rabbit@iZq8zm58oi8hwrZ-mnesia",
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>                                50000000,undefined,100,10000,undefined,false,
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>                                false,9,120000}
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1> ** Reason for termination ==
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1> ** {could_not_determine_disk_free,
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>        [{rabbit_disk_monitor,get_disk_free,2,
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>             [{file,"rabbit_disk_monitor.erl"},{line,254}]},
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>         {rabbit_disk_monitor,internal_update,1,
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>             [{file,"rabbit_disk_monitor.erl"},{line,216}]},
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>         {rabbit_disk_monitor,enable,1,
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>             [{file,"rabbit_disk_monitor.erl"},{line,355}]},
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>         {rabbit_disk_monitor,handle_info,2,
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>             [{file,"rabbit_disk_monitor.erl"},{line,164}]},
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>         {gen_server,try_dispatch,4,[{file,"gen_server.erl"},{line,1120}]},
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>         {gen_server,handle_msg,6,[{file,"gen_server.erl"},{line,1197}]},
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>         {proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,240}]}]}
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1> 
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>   crasher:
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>     initial call: rabbit_disk_monitor:init/1
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>     pid: <0.5482.1>
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>     registered_name: rabbit_disk_monitor
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>     exception exit: could_not_determine_disk_free
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>       in function  rabbit_disk_monitor:get_disk_free/2 (rabbit_disk_monitor.erl, line 254)
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>       in call from rabbit_disk_monitor:internal_update/1 (rabbit_disk_monitor.erl, line 216)
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>       in call from rabbit_disk_monitor:enable/1 (rabbit_disk_monitor.erl, line 355)
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>       in call from rabbit_disk_monitor:handle_info/2 (rabbit_disk_monitor.erl, line 164)
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>       in call from gen_server:try_dispatch/4 (gen_server.erl, line 1120)
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>       in call from gen_server:handle_msg/6 (gen_server.erl, line 1197)
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>     ancestors: [rabbit_disk_monitor_sup,rabbit_sup,<0.219.0>]
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>     message_queue_len: 0
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>     messages: []
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>     links: [<0.354.0>]
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>     dictionary: []
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>     trap_exit: false
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>     status: running
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>     heap_size: 6772
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>     stack_size: 28
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>     reductions: 33064
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1>   neighbours:
2022-08-16 12:26:52.084000+08:00 [error] <0.5482.1> 
2022-08-16 12:26:52.084000+08:00 [error] <0.354.0>     supervisor: {local,rabbit_disk_monitor_sup}
2022-08-16 12:26:52.084000+08:00 [error] <0.354.0>     errorContext: child_terminated
2022-08-16 12:26:52.084000+08:00 [error] <0.354.0>     reason: could_not_determine_disk_free
2022-08-16 12:26:52.084000+08:00 [error] <0.354.0>     offender: [{pid,<0.5482.1>},
2022-08-16 12:26:52.084000+08:00 [error] <0.354.0>                {id,rabbit_disk_monitor},
2022-08-16 12:26:52.084000+08:00 [error] <0.354.0>                {mfargs,{rabbit_disk_monitor,start_link,[50000000]}},
2022-08-16 12:26:52.084000+08:00 [error] <0.354.0>                {restart_type,{transient,1}},
2022-08-16 12:26:52.084000+08:00 [error] <0.354.0>                {shutdown,300000},
2022-08-16 12:26:52.084000+08:00 [error] <0.354.0>                {child_type,worker}]

**Comment by michaelklishin** [comment]
@noobwu this was addressed by https://github.com/rabbitmq/rabbitmq-server/pull/4328 and https://github.com/rabbitmq/rabbitmq-server/pull/5241. Both [shipped in `3.10.7`](}}https://github.com/rabbitmq/rabbitmq-server/releases/tag/v3.10.7).
